### PR TITLE
Utility functions for using dynamic Kubelet configuration from a test

### PR DESCRIPTION
**Pull request by mtaufen**
/cc @vishh @dchen1107

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32530)

### CODE REVIEW

**Comment by Random-Liu** [comment]
@mtaufen Is this WIP or ready for review? :)

**Comment by mtaufen** [comment]
@Random-Liu Just exited WIP. Good for review now.

**Comment by Random-Liu** [comment]
@k8s-bot test this issue #IGNORE.

**Comment by mtaufen** [comment]
@random-liu this looks good to go PTAL

File: test/e2e_node/util.go [hunk 1]
```diff
@@ -60,3 +70,155 @@ func getNodeSummary() (*stats.Summary, error) {
...
+}
+
+func getCurrentKubeletConfigMap(f *framework.Framework) (*api.ConfigMap, error) {
+	return f.ClientSet.Core().ConfigMaps("kube-system").Get(fmt.Sprintf("kubelet-%s", framework.TestContext.NodeName))
+}
+
+// Creates or updates the configmap for KubeletConfiguration, waits for the Kubelet to restart
+// with the new configuration. Returns an error if the configuration after waiting 40 seconds
+// doesn't match what you attempted to set.
+func setKubeletConfiguration(f *framework.Framework, kubeCfg *componentconfig.KubeletConfiguration) error {
```
Comment (Random-Liu) [inline test/e2e_node/util.go]: Can we check the feature gates first and only set kubelet configuration when dynamic kubelet configuration is enabled?

File: test/e2e_node/util.go [hunk 2]
```diff
@@ -60,3 +70,155 @@ func getNodeSummary() (*stats.Summary, error) {
...
+	const (
+		restartGap = 30 * time.Second
+	)
+
+	// Check whether a configmap for KubeletConfiguration already exists
+	_, err := getCurrentKubeletConfigMap(f)
+	if fmt.Sprintf("%v", err) == fmt.Sprintf("configmaps \"kubelet-%s\" not found", framework.TestContext.NodeName) {
```
Comment (Random-Liu) [inline test/e2e_node/util.go]: Should use `k8s.io/kubernetes/pkg/api/errors.IsNotFound`

File: test/e2e_node/util.go [hunk 3]
```diff
@@ -60,3 +70,155 @@ func getNodeSummary() (*stats.Summary, error) {
...
+
+	// Retrieve the new config and compare it to the one we attempted to set
+	newKubeCfg, err := getCurrentKubeletConfig()
+	if err != nil {
+		return err
+	}
+
+	// Return an error if the desired config is not in use by now
+	if !reflect.DeepEqual(*kubeCfg, *newKubeCfg) {
+		return fmt.Errorf("Either the Kubelet did not restart or it did not present the modified configuration via /configz after restarting.")
```
Comment (Random-Liu) [inline test/e2e_node/util.go]: nit: s/Either/either

File: test/e2e_node/util.go [hunk 4]
```diff
@@ -60,3 +70,155 @@ func getNodeSummary() (*stats.Summary, error) {
...
+		if resp.StatusCode != 200 {
+			glog.Errorf("/configz response status not 200, retrying. Response was: %+v", resp)
+			return false
+		}
+		return true
+	}, timeout, pollInterval).Should(Equal(true))
+	return resp
+}
+
+// Decodes the http response  from /configz and returns a componentconfig.KubeletConfiguration (internal type).
```
Comment (Random-Liu) [inline test/e2e_node/util.go]: nit: remove extra space.

File: test/e2e_node/util.go [hunk 5]
```diff
@@ -60,3 +70,155 @@ func getNodeSummary() (*stats.Summary, error) {
...
+	err = api.Scheme.Convert(&configz.ComponentConfig, &kubeCfg, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &kubeCfg, nil
+}
+
+// Uses KubeletConfiguration to create a `kubelet-<node-name>` ConfigMap in the "kube-system" namespace.
+func createConfigMap(f *framework.Framework, kubeCfg *componentconfig.KubeletConfiguration) (*api.ConfigMap, error) {
```
Comment (Random-Liu) [inline test/e2e_node/util.go]: Most code in `createConfigMap` and `updateConfigMap` is duplicated. Can we unify these 2 functions?

**Review by Random-Liu — COMMENTED** [review]
One more nit: These functions are supper useful, can we move them into framework?

**Comment by Random-Liu** [comment]
LGTM overall with nits.

File: test/e2e_node/util.go (hunk 1 above)
Comment (mtaufen) [inline test/e2e_node/util.go]: Left a TODO pending #31556

File: test/e2e_node/util.go (hunk 2 above)
Comment (mtaufen) [inline test/e2e_node/util.go]: Done

File: test/e2e_node/util.go (hunk 3 above)
Comment (mtaufen) [inline test/e2e_node/util.go]: Done

File: test/e2e_node/util.go (hunk 4 above)
Comment (mtaufen) [inline test/e2e_node/util.go]: Done

File: test/e2e_node/util.go (hunk 5 above)
Comment (mtaufen) [inline test/e2e_node/util.go]: Done

**Comment by mtaufen** [comment]
WRT moving them into `test/e2e/framework`, these utils currently assume that tests are running on the same node as the pods; to extend to general e2e tests we would need to allow utils to set configuration for other nodes in the cluster as well. For now I think there is enough value in having these utils available for node e2e tests that we can add the cluster test capability later.

**Comment by mtaufen** [comment]
@k8s-bot test this

File: cluster/gce/gci/node-helper.sh [hunk 6]
```diff
@@ -1,3 +1,4 @@
+dockerdockerpkgmountflags
```
Comment (Random-Liu) [inline cluster/gce/gci/node-helper.sh]: ??? Something wrong here?

Comment (mtaufen) [inline cluster/gce/gci/node-helper.sh]: uuh that's weird, wonder where that came from...

**Comment by mtaufen** [comment]
@k8s-bot cvm gke e2e test this

**Comment by mtaufen** [comment]
@Random-Liu The necessary FeatureGates PR went in, so I added the check for whether dynamic config is enabled to the set function. PTAL.

**Comment by Random-Liu** [comment]
LGTM. This is quite useful! Many changes in my mind on test framework needs this feature!

**Pull request merged**